### PR TITLE
feat: Add show derivation

### PR DESCRIPTION
**Pull request by boris-marinov**
I took the code from the 'Either' module and put it into a generic derivation.

Added a representation for the constructor functions (All other values have one, so I think it makes sense).

Using `getOwnPropertyNames` for listing the properties for a given value, because according to [section 14.4.2 from this source](http://exploringjs.com/es6/ch_oop-besides-classes.html) it returns the values in a deterministic order. According to the [MDN docs it doesn't](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyNames).  In practice both `getOwnPropertyNames` and `Object.keys` work fine. Not so important  since AFAIK names are just for debugging purpose.  We may event consider switching to `keys` to prevent the displaying of non-enumerable properties. What do you think?

### CODE REVIEW

File: src/core/adt/show.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+const { tagSymbol, typeSymbol } = require('./core');
+
+const displayValues = (object) => Object.getOwnPropertyNames(object)
```
Comment (robotlolita) [inline src/core/adt/show.js]: We should use a `key: value` pair serialisation here, similar to what JSON.stringify does. This avoids relying on key order (which is only guaranteed to be insertion order in ES2015), and helps people with what they should be destructuring in the `.cata()` method. I think `Object.keys` makes more sense, we should clarify somewhere that the ADT module and its derivations only consider enumerable keys. Something like:

```js
const displayValues = (object) =>
  Object.keys(object)
        .map((key) => [`${JSON.stringify(key)}: ${JSON.stringify(String(value))}`])
        .join(', ')
```

**Comment by robotlolita** [comment]
I guess the representation for the root type is okay. We might want to throw an error if someone gives an empty object to `data()`, because `data('type', {}).derive(show)` wouldn't call the derivation function, and it's also most likely a mistake (since it doesn't make much sense to not have any variant).

**Comment by boris-marinov** [comment]
>We probably want to handle values in a better way, passing primitives to JSON.stringify (so we don't have problems with strings), and just using String(value) otherwise.

How do we deal with plain objects? Meaning how do we know which ones are plain. The ones that don't have a prototype?

The `isPlainObject` function from lodash is [long and weird](https://github.com/lodash/lodash/blob/master/dist/lodash.js#L11274).

**Comment by robotlolita** [comment]
I was thinking more along the lines of:

```js
function isPrimitive(x) {
  return typeof x === 'string' 
  ||     typeof x === 'number' 
  ||     typeof x === 'boolean';
}

function showValue(x) {
  return x === null        ?  'null'
  :      x === undefined   ?  'undefined'
  :      Array.isArray(x)  ?  `[${x.map(showValue).join(', ')}]`
  :      isPrimitive(x)    ?  JSON.stringify(x)
  :      /* otherwise */      String(x) // calls .toString()
}
```

This might give people `value: [object Object]` when they have a value in the variant that doesn't implement a custom .toString(), but I think at that point they'd want to write a custom .toString() method anyway, since there isn't a simple generic way of describing these objects in general.

**Comment by boris-marinov** [comment]
I think we still can and should handle the presentation of plain objects in the same way that you handle arrays in your example. Especially since we are overwriting the node.js console representation which handles them by default.

The check whether something is a plain object can be hard, but I think I've come up with a solution  - instead of inspecting the object itself we can just check if its `toString` function is the default one.

I am not sure the referential equality would work in all cases but that can be tested.

File: src/core/adt/show.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+const objectToString = (object) => {
+  return object === null                                ?  nullToString
+  :      object.toString === Array.prototype.toString   ?  arrayToString
+  :      object.toString === Object.prototype.toString  ?  plainObjectToString
+  :      /* otherwise */                                   object.toString;
+};
+
+const showValue = (value) =>
+  typeof value !== 'object' ? JSON.stringify(value) : objectToString(value).call(value);
```
Comment (robotlolita) [inline src/core/adt/show.js]: We need to handle `undefined` here too, since JSON.stringify doesn't handle it :/

**Comment by robotlolita** [comment]
Oh, I think this is a good direction. :+1:

**Comment by boris-marinov** [comment]
It is. 

It does not work across realms though:
```
win = window.open('https://github.com')
> Window {external: Object, chrome: Object, document: document, speechSynthesis: SpeechSynthesis, caches: CacheStorage…}
win.Object.toString === Object.toString
> false
```
But I am not even sure if it is worth it to add code that deals with this case. We can do: 
```javascript
const hasCustomToString = (object) => {
  while (Object.getPrototypeOf(object) !== null) {
    if (object.hasOwnProperty('toString')) {
      return true
    }
    object = Object.getPrototypeOf(object)
  }
  return false
}

console.assert(hasCustomToString({}) === false)
console.assert(hasCustomToString({toString(){}}) === true)
console.assert(hasCustomToString(Object.create( { toString(){} }) ) === true)

```
I don't know...

**Comment by boris-marinov** [comment]
Regarding the throwing of error on invalid input: Yes, we should definitely be doing this for all methods that we provide (unless performance).  Maybe folktale validations are a good fit for this problem?

**Comment by robotlolita** [comment]
Ah, I hadn't thought about the realms stuff. That makes this complicated.

We could use `Array.isArray(value)` and `value.toString() === '[obect Object]'` to check for that instead, which works cross-realm. This ignores that people would change the Array's `.toString` representation, but I think that's an okay compromise.

**Comment by boris-marinov** [comment]
Indeed it is. 
Fixed.
The funny thing is that the `undefined` handling was working since `JSON.stringify(undefined)` returns `undefined`. Added a check though.

**Comment by robotlolita** [comment]
Great, thanks :)

The problem with `undefined` is that its conversion to strings isn't consistent in the standard library, like `String(undefined)` is `"undefined"`, but `[undefined].toString()` is `""` .-.

**Pull request merged**